### PR TITLE
feature: Market Mode UI - Main Page + Sell Produce Page

### PR DESCRIPTION
**Pull request by neelp1030**
https://streamable.com/ha9dx2 - Demo Video (I may have missed a few things to touch upon so please let me know and I can showcase)

Some things to keep in mind:

1) The contents of the main Market page are currently not what we intend on having in the final version. Right now, I have simply reused the markets + produce quota progress component that we use for Home -> Weekly Quotas page; technically this is valuable content even for the market page but is redundant nevertheless. For the final version, I am planning on replacing the contents of that page with a pie graph where each section represents a market, and the size of the section is based on the total earnings of that market. The total earnings will be computed using transactions info, and I am working on this currently.

2) Speech-to-text is not implemented for Sell Produce page yet. I can work on this as well, probably with @SathursanS's help, but maybe we should reconsider if there is value in this feature for market.

3) For the sell produce page, I have made it so that if a quota does not exist for a specific produce, then we do not display the progress bar, just wanted to make it clear.

4) For the sell produce page, the + and - buttons are respectively disabled when the limits are reached (lower limit of 0 and upper limit of the inventory for that produce).

### CODE REVIEW

**Review by adshayanB — APPROVED** [review]

**Comment by RishanR** [comment]
The demo is looking awesome! Great work putting all this together! Before I took a look at the PR, I have a few comments based on the demo:

1. The markets page seems to be identical to the quotas page at the moment. I don't think we should show the quotas here. It should just be the market names and any other market related information. At the moment I think its just market name so I can see how it may look weird since we don't show much info there. But for now, just make it so that each card is just the market name.
2. In the main markets page, there should be an "+" button that allows the user to add a new market. This will navigate to an Add Market screen where users can add a new market. When the user then submits the new market form, a new market will be added to the database.
3. I think we should somehow indicate to the user that the green bars under every incrementer in the specific market page represents quota progress, since it may be a little unclear. Maybe instead of just writing "1/10" in the bar, we can write "Quota Progress: 1/10".
4. To make the UI consistent and easily readable, we should also add bars to produce that don't have quotas. For example, in your demo, there is a quota bar under orange, but no bar under apple since there's no quota associated with it. I think instead of no bar, we should show an empty bar with the message "No Quota Available" inside of it.

**Comment by neelp1030** [comment]
> The demo is looking awesome! Great work putting all this together! Before I took a look at the PR, I have a few comments based on the demo:
> 
> 1. The markets page seems to be identical to the quotas page at the moment. I don't think we should show the quotas here. It should just be the market names and any other market related information. At the moment I think its just market name so I can see how it may look weird since we don't show much info there. But for now, just make it so that each card is just the market name.
> 2. In the main markets page, there should be an "+" button that allows the user to add a new market. This will navigate to an Add Market screen where users can add a new market. When the user then submits the new market form, a new market will be added to the database.
> 3. I think we should somehow indicate to the user that the green bars under every incrementer in the specific market page represents quota progress, since it may be a little unclear. Maybe instead of just writing "1/10" in the bar, we can write "Quota Progress: 1/10".
> 4. To make the UI consistent and easily readable, we should also add bars to produce that don't have quotas. For example, in your demo, there is a quota bar under orange, but no bar under apple since there's no quota associated with it. I think instead of no bar, we should show an empty bar with the message "No Quota Available" inside of it.

Points 3 and 4 have been addressed! For point 1, I addressed the current state of the main market page in the description of this PR; I'm planning on changing it to be a pie chart based on total earnings per market, but probably it's better to do this work in a separate PR as this is big enough. I think I will just make it have the market names for this PR, and then change it to incorporate the pie chart in a new PR. Point 2 is an excellent observation, which I'm working on right now.

**Comment by neelp1030** [comment]
@RishanR I have addressed the requirement of the add/edit market functionality. Please take a look when you get some free time :) Here's a video demonstrating this new functionality: https://streamable.com/qjt66y

File: app/src/main/java/com/example/farmeraid/data/MarketRepository.kt [hunk 1]
```diff
@@ -13,9 +16,46 @@ import kotlinx.coroutines.tasks.await
...
+                    )
+                }
+
+                ResponseModel.FAResponse.Success
+            } catch(e: Exception) {
+                return ResponseModel.FAResponse.Error(e.message?:"Error creating a market. Please try again.")
+            }
+        } else {
+            return try {
+                db.collection("market").document(querySnapshot.documents[0].id).update(
```
Comment (adshayanB) [inline app/src/main/java/com/example/farmeraid/data/MarketRepository.kt:49]: Nice adding it under one function

**Comment by neelp1030** [comment]
@RishanR I've modified the main market page as per your request (point #1); it now displays the produce and associated prices in the card. Here's what it looks like: 

image

Let me know any feedback on the information we show here. I think this might be good to actually leave as a final result for the main market page, as it is in line with the quotas page where we also just show the quota information in the cards. If we choose to do this, that would mean that we either drop the task of creating some sort of useful graph regarding the markets, or create a new page and put it there. What do you guys think?

**Comment by neelp1030** [comment]
Here's a preview of the new Add/Edit Market Page after the changes to allow for decimal numbers to be entered for the price:

**Comment by RishanR** [comment]
Took a look at the new video you sent as well as the new screenshots for main markets and add/edit market! It's looking amazing!! Good idea with showing the produce and their associated prices in the main markets page!

Minor nits:
- In the main markets page and sell produce page, mention what the price next to the produce means (is it price per produce or total produce earnings for that market. For example, you can add a little "$20/produce" or something like that if its for per produce.
- Add some more spacing between the each produce item in the Sell Produce page for added readability (I've mentioned where you can add this in the code in the comments below)
- Make all the prices have 2 decimal places after the decimal point

Functionality wise this is looking great! Amazing work :)

File: app/src/main/java/com/example/farmeraid/market/sell_produce/views/SellProduceView.kt [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+                                    expectedProgress = if (produceSell.produceQuotaTotalGoal != -1) (produceSell.produceQuotaCurrentProgress.toFloat() + produceSell.produceCount) / produceSell.produceQuotaTotalGoal else 0f,
+                                    fontSize = 14.sp,
+                                    containerColor = PrimaryColour.copy(alpha = 0.2f),
+                                    progressColor = PrimaryColour,
+                                ),
+                                showProgressBar = true,
+                            ),
+                            modifier = Modifier.fillMaxWidth(),
+                        )
+                        Spacer(modifier = Modifier.height(10.dp))
```
Comment (RishanR) [inline app/src/main/java/com/example/farmeraid/market/sell_produce/views/SellProduceView.kt:154]: Make this 20.dp

**Review by RishanR — APPROVED** [review]
Looking great! Well done :)

**Pull request merged**